### DISCUSSION TITLE
Abstraction methods of PdfViewerController.goTo

### DISCUSSION THREAD

**Discussion by xaqtr** [espresso3389/flutter_pdf_render#71] [Ideas]
Hi,

first of all thanks for this library @espresso3389, it's really saving us for a work related project right now due to its flexiblity and features.

As I was trying to integrate this plugin into my App, I was kind of lost on how to programmatically scale into the PDF with the `PdfViewerController` (I wanted to implement a double tap zoom in /out functionality).

I digged into the plugin code and realized that you have the control about everything by using the `PdfViewerController.goTo` method. and passing a suited `Matrix4` value. So my question is whether it would be possible to have methods that abstract the usage of this `Matrix4`. I think this would benefit many developers as in my opinion it requires some additional knowledge hurdle.

So my proposal would be some additional methods for `PdfViewerController`, e.g.:

```
// In this case, scale would be the destination scale 
// (so calling this method more than once with the same parameter will not change the scaling / position)
Future<void> scaleTo({ required double scale, Duration duration })

// Pan to a relative point of the specified page
// x and y would be relative values of the page, where the destination origin will be
// e.g.: pageNumber = 1, x = 0.5, y = 0.5 will put the origin into the center of the first page
Future<void> panTo({ required int pageNumber, required double x, required double y, Duration duration })
```

I dont' have the best understand about the usage of `Matrix4` yet, so I'm not sure whether it's even possible to have such methods, but I would like to hear some opinions on that.
In the meantime I will try to implement the above mentioned methods.

**Comment by espresso3389** [comment]
I see your point. They seems very convenient when implementing typical PDF viewer.
OK, I'll implement them.

**Comment by espresso3389** [comment]
daa6d7d087ea0ab454161783c3879e2e0e9e35ba implements [panTo](https://github.com/espresso3389/flutter_pdf_render/blob/daa6d7d087ea0ab454161783c3879e2e0e9e35ba/lib/pdf_render_widgets.dart#L592) method:

```dart
/// Pan to the specified page point.
///
/// [x],[y] should be in [0 1] range and they indicate relative position in the page:
/// - 0 for top/left
/// - 1 for bottom/right
/// - 0.5 for center (the default)
///
/// [anchor] specifies which view corner, edge, or center the point specified by ([x],[y]) is anchored to.
///
/// [zoomRatio] specifies the zoom ratio. The default is to use the zoom ratio that fit the page into the view.
/// If you want to keep the current zoom ratio, use [PdfViewerController.zoomRatio] for the value.
///
/// If the page does not exist in the layout, it returns null.
/// If the controller is not ready([isReady]), the method throws an exception.
Future<void> panTo({
  required int pageNumber,
  double? padding,
  double x = 0.5,
  double y = 0.5,
  PdfViewerAnchor anchor = PdfViewerAnchor.center,
  double? zoomRatio,
  Duration duration = const Duration(milliseconds: 500),
)
```

**Reply by espresso3389** [reply]
- `scale` is renamed to `zoomRatio` in correspondence to the existing `zoomRatio` property.
- `anchor` is where the point of `(x,y)` is anchored to. Based on your original use case, it's default is set to `center`. But you can of course change it to one of `topLeft`, `top`, `topRight`, `left`, `center`, `right`, `bottomLeft`, `bottom`, `bottomRight`.

**Comment by espresso3389** [comment]
9b57f1ec3390f36a9cdb6ba4a97c9144f0362e37:
- `panTo` is renamed to [goToPointInPage](https://github.com/espresso3389/flutter_pdf_render/blob/9b57f1ec3390f36a9cdb6ba4a97c9144f0362e37/lib/pdf_render_widgets.dart#L592)
- [setZoomRatio](https://github.com/espresso3389/flutter_pdf_render/blob/9b57f1ec3390f36a9cdb6ba4a97c9144f0362e37/lib/pdf_render_widgets.dart#L634) is introduced

```dart
final controller = PdfViewerController();
TapDownDetails? doubleTapDetails;

...

GestureDetector(
  // Supporting double-tap gesture on the viewer.
  onDoubleTapDown: (details) => doubleTapDetails = details,
  onDoubleTap: () => controller.ready?.setZoomRatio(
    zoomRatio: controller.zoomRatio * 1.5,
    center: doubleTapDetails!.localPosition,
  ),
  child: PdfViewer.openAsset(
    'assets/hello.pdf',
    viewerController: controller,
    ...
```

**Comment by espresso3389** [comment]
[1.0.22](https://pub.dev/packages/pdf_render/versions/1.0.22) that contains the changes is released!

**Comment by espresso3389** [comment]
Because of the consistency/compatibility to the existing codes/interfaces, the naming is changed from your first suggestion, I think the designs of the new methods covers what you want to do.

Anyway, please test them and feel free to suggest/comment anything here!

**Reply by xaqtr** [reply]
Thanks a lot! I actually will start integrating both new methods in the upcoming week for a production app and will let you know, if I see something odd.
